src/collapse/dtrace.rs

```diff
@@ -0,0 +1,219 @@
+use super::Collapse;
+use smallvec::SmallVec;
+use std::collections::{HashMap, VecDeque};
+use std::io;
+use std::io::prelude::*;
+
+/// Settings that change how frames are named from the incoming stack traces.
+///
+/// All options default to off.
+#[derive(Clone, Debug, Default)]
+pub struct Options {
+    /// Include raw addresses (e.g., `0xbfff0836`) where symbols can't be found.
+    pub include_addrs: bool,
+
+    /// Annotate kernel functions with a `_[k]` suffix.
+    pub annotate_kernel: bool,
+}
+
+/// A stack collapser for the output of `perf script`.
+///
+/// To construct one, either use `perf::Folder::default()` or create an [`Options`] and use
+/// `perf::Folder::from(options)`.
+#[derive(Default)]
+pub struct Folder {
+    /// All lines until the next empty line are stack lines.
+    in_event: bool,
+
+    /// Skip all stack lines in this event.
+    skip_stack: bool,
+
+    /// Function entries on the stack in this entry thus far.
+    stack: VecDeque<String>,
+    count: usize,
+    /// Number of times each call stack has been seen.
+    occurrences: HashMap<String, usize>,
+
+    /// Current comm name.
+    ///
+    /// Called pname after original stackcollapse-perf source.
+    pname: Option<String>,
+
+    //    event_filtering: EventFilterState,
+    opt: Options,
+}
+
+impl Collapse for Folder {
+    fn collapse<R, W>(&mut self, mut reader: R, writer: W) -> io::Result<()>
+    where
+        R: BufRead,
+        W: Write,
+    {
+        let mut line = String::new();
+        //TODO: clean this up, we skip the first 4 lines:
+        //        CPU     ID                    FUNCTION:NAME
+        reader.read_line(&mut line)?;
+        //        0  64091                        :tick-60s
+        reader.read_line(&mut line)?;
+        //
+        reader.read_line(&mut line)?;
+        //
+        reader.read_line(&mut line)?;
+        loop {
+            line.clear();
+
+            if reader.read_line(&mut line)? == 0 {
+                break;
+            }
+
+            if line.starts_with('#') {
+                continue;
+            }
+
+            let line = line.trim_end();
+            if line.is_empty() {
+                self.after_event();
+            } else {
+                self.on_line(line);
+            }
```

Comment:
I think I'd change this slightly since the DTrace format seems to be:
```
stack line
stack line
count

stack line
stack line
count
```
I'd do:
```rust
if line.is_empty() {
    continue;
} else if let Ok(count) = line.parse::<usize>() {
    self.on_stack_end(count);
} else {
    self.on_stack_line(line);
}
```
That way you also don't need to remember how many lines to skip at the head of the file. Just ignore any lines up to the first empty line, and from that on do normal processing. Any and all empty lines will just be skipped.

Reply:
that would lead to some warnings about odd stack lines at the beginning.

Reply:
Ah, no, I'm proposing that you _first_ skip leading non-empty lines (and don't warn for them). Only once you've hit the first empty line do you start "normal" line processing (including warning for malformed lines).

Reply:
ahhh that makes more sense!